lib/blocks.php

```diff
@@ -102,28 +102,48 @@ function do_blocks( $content ) {
 	global $wp_registered_blocks;
 
 	// Extract the blocks from the post content.
-	$matcher = '/<!--\s*wp:([a-z](?:[a-z0-9\/]+)*)\s+((?:(?!-->).)*)\s*\/?-->(?:.*?<!--\s*\/wp:\g1\s+-->)?/s';
+	$matcher = '#' . join( '', array(
+		'(?P<opener><!--\s*',
+		'wp:(?P<block_name>[a-z](?:[a-z0-9/]+)*)\s+',
+		'(?P<attributes>(?:(?!-->).)*)',
+		'\s*/?-->\n?)',
+		'(?:',
+		'(?P<content>.*?)',
+		'(?P<closer><!--\s*/wp:\g{block_name}\s+-->\n?)',
+		')?',
+	) ) . '#s';
 	preg_match_all( $matcher, $content, $matches, PREG_OFFSET_CAPTURE );
 
 	$new_content = $content;
+	$offset_differential = 0;
```

Comment:
Curious if it would be simpler to implement as `while ( preg_match( $matcher, $content, $match ) ) {`

Maybe a performance hit by repeatedly matching the same string.